spatialmath/capsule.go

```diff
@@ -0,0 +1,328 @@
+package spatialmath
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+
+	"github.com/golang/geo/r3"
+	commonpb "go.viam.com/api/common/v1"
+
+	"go.viam.com/rdk/utils"
+)
+
+// capsule is a collision geometry that represents a capsule, it has a pose and a radius that fully define it.
+//
+//	 __________________
+//	/                  \
+//
+// x|  |--------------|  |x
+//
+//	\__________________/
+//
+// Length is the distance between the x's, or internal segment length + 2*radius.
+type capsule struct {
+	// this is the pose of one end of the capsule. The full capsule extends `length` mm outwards in the direction of
+	// the pose's orientation
+	pose   Pose
+	radius float64
+	length float64 // total length of the capsule, tip to tip
+	label  string
+
+	// These values are generated at geometry creation time and should not be altered by hand
+	// They are stoed here because they are useful and expensive to calculate
+	segA r3.Vector // Proximal endpoint of capsule line segment. First point from `pose` to be surrounded by `radius` of capsule
+	segB r3.Vector // Distal endpoint of capsule line segment. Most distal point to be surrounded by `radius` of capsule
+}
+
+// NewCapsule instantiates a new capsule Geometry.
+func NewCapsule(offset Pose, radius, length float64, label string) (Geometry, error) {
+	if radius < 0 {
+		return nil, newBadGeometryDimensionsError(&capsule{})
+	}
+	if length == radius*2 || length == 0 {
```

Comment:
The convenience here is allowing spheres and capsules to be e.g. inferred from proto with the same code.

If we allow `0 <= length <= 2*r`, then the user can pass in potentially meaningful values that will be silently ignored. `0` is the only value in that range where the user is getting exactly what they asked for.

Reply:
I don't follow this.  Proto geometries are strongly typed so no need for inference.  Config geometry is inferred but if any L is specified we know we can be using a capsule.  Can you give an example of how this would be broken?

Reply:
I was referring to configs, not raw protos, apologies; I misspoke. The ode in question is here:

https://github.com/viamrobotics/rdk/blob/140edb771726c511cecd4e9b68a7f2c78115eff4/spatialmath/geometry.go#L105

If L is zero, the we return a sphere, which is correct.

If L is 2, but radius is 1.5, then that's an invalid capsule, as length would have to be greater than 3. We shouldn't coerce to a sphere in this case, because it's unknown if the user intended a capsule of length 5, or a sphere of diameter 3.

In that case we should error to the user.

Reply:
Ah ok I follow now.  I'm not sure I entirely agree that reporting that as an error is the best course of action, but I can get behind it.

Reply:
If you prefer I can make it so that length == 0 will also be an error; that way we're never assuming the user is shortcutting.

Reply:
I think you're right on `ParseConfig`. Was trying to save some code, but that does leave an ambiguity here that would be corrected.

Length == 0 will become an error for capsules. `ParseConfig` will look at L, and call `NewSphere` if L == 0

Reply:
Updated